Disable sharing calendars via link when sharing via link is disabled

pboguslawski opened the pull request:
This mod disallows sharing calendars via link when `shareapi_allow_links` is disabled.

Related: https://github.com/nextcloud/calendar/issues/525
Related: https://github.com/nextcloud/calendar/issues/4399
Author-Change-Id: IB#1126264
Signed-off-by: Pawel Boguslawski <pawel.boguslawski@ib.pl>

pboguslawski committed:
Disable sharing calendars via link when sharik via link is disabled

This mod disallows sharing calendars via link when `shareapi_allow_links`
is disabled.

Related: https://github.com/nextcloud/calendar/issues/525
Related: https://github.com/nextcloud/calendar/issues/4399
Author-Change-Id: IB#1126264
Signed-off-by: Pawel Boguslawski <pawel.boguslawski@ib.pl>

miaulalala commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -570,6 +575,10 @@ public function getPublicCalendars() {
 	public function getPublicCalendar($uri) {
+		// Sharing via link must be enabled.
+		if ($this->config->getAppValue('core', 'shareapi_allow_links', 'yes') === 'no' ) {
+			throw new NotFound('Node with name \'' . $uri . '\' could not be found');
```
Please use an exception that reflects what is happening: `\Sabre\DAV\Exception\Forbidden` would be a better exception to use here.

pboguslawski replied:
Fixed in bfd8c622d99f24d7769db39b095ce00176e6da8a.

miaulalala commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -512,6 +513,10 @@ private function getUserDisplayName($uid) {
 	 * @return array
 	 */
 	public function getPublicCalendars() {
+		// Sharing via link must be enabled.
```
```suggestion
		// Sharing via link is allowed by default, but if the option is set it should be checked.
```

miaulalala commented:
Thanks for your PR!  A couple small change requests only :)

tcitworld commented on apps/dav/lib/CalDAV/Publishing/PublishPlugin.php:
```diff
@@ -205,6 +206,11 @@ public function httpPost(RequestInterface $request, ResponseInterface $response)
 				}
 			}
 
+			// Sharing via link must be enabled.
```
It might be better to just avoid registering the SabreDAV plugin altogether when the feature is disabled:
https://github.com/nextcloud/server/blob/905753be42af7e8c5af304b0d3c93f5d9bcea789/apps/dav/lib/Server.php#L184-L187

Otherwise you would need to avoid exposing the plugin features as well.
https://github.com/nextcloud/server/blob/905753be42af7e8c5af304b0d3c93f5d9bcea789/apps/dav/lib/CalDAV/Publishing/PublishPlugin.php#L86-L89
https://github.com/nextcloud/server/blob/905753be42af7e8c5af304b0d3c93f5d9bcea789/apps/dav/lib/CalDAV/Publishing/PublishPlugin.php#L138

pboguslawski replied:
Fixed in bfd8c622d99f24d7769db39b095ce00176e6da8a.

miaulalala commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -512,6 +513,10 @@ private function getUserDisplayName($uid) {
 	 * @return array
 	 */
 	public function getPublicCalendars() {
```
Let's deprecate it then please @tcitworld

tcitworld replied:
Well as it's still being used in the tests we can refactor them remove it altogether. Even though the method is public it's kind of a private API, and I don't seem to find anywhere it's been used anyway.

pboguslawski replied:
Fixed in bfd8c622d99f24d7769db39b095ce00176e6da8a.

pboguslawski committed:
Disable sharing calendars via link adjusted

Fixes: 45eefc221d4439c99fd2c8badfe93dc3af40f028
Related: https://github.com/nextcloud/server/pull/34372
Author-Change-Id: IB#1126264
Signed-off-by: Pawel Boguslawski <pawel.boguslawski@ib.pl>

pboguslawski commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -570,6 +575,10 @@ public function getPublicCalendars() {
 	 */
 	public function getPublicCalendar($uri) {
+		// Sharing via link must be enabled.
+		if ($this->config->getAppValue('core', 'shareapi_allow_links', 'yes') === 'no' ) {
```
Fixed in bfd8c622d99f24d7769db39b095ce00176e6da8a.

pboguslawski commented:
Current solution (unregistering plugin) does [not work correctly](https://github.com/nextcloud/calendar/pull/4572#issuecomment-1295231692) so please hold with merging until final solution is proposed.

tcitworld commented:
@pboguslawski That's correct, thanks for checking! Sadly the AllowedSharingModes didn't fallback as I planned when the plugin was unregistered, so I made #34873 to extract exposing this information into it's own plugin. That should do it, please have a go and test.

pboguslawski commented:
Replaced with https://github.com/nextcloud/server/pull/34873.
